File: core/models/message.py
"""Message models."""

from typing import Any, Literal

from pydantic import BaseModel

from .message_time import MessageTime
from .model_info import ModelInfo
from .path_info import PathInfo
from .token_info import TokenInfo


class UserMessage(BaseModel):
    id: str
    sessionID: str
    role: Literal["user"] = "user"
    time: MessageTime
    agent: str
    model: ModelInfo
    system: str | None = None
    tools: dict[str, bool] | None = None


class AssistantMessage(BaseModel):
    id: str
    sessionID: str
    role: Literal["assistant"] = "assistant"
    time: MessageTime
    parentID: str
    modelID: str
    providerID: str
    mode: str
    path: PathInfo
    cost: float
    tokens: TokenInfo
    finish: str | None = None
    summary: bool | None = None
    error: dict[str, Any] | None = None


Message = UserMessage | AssistantMessage

File: core/models/message.ts
/**
 * Message models - user and assistant messages in a session.
 */

export type MessageStatus = 'pending' | 'streaming' | 'completed' | 'failed' | 'aborted';

export interface MessageTime {
  created: number;
  completed?: number;
}

export interface ModelInfo {
  providerID: string;
  modelID: string;
}

export interface PathInfo {
  cwd: string;
  root: string;
}

export interface TokenInfo {
  input: number;
  output: number;
  reasoning: number;
  cache?: {
    read: number;
    write: number;
  };
}

export interface UserMessage {
  id: string;
  sessionID: string;
  role: 'user';
  time: MessageTime;
  status: MessageStatus;
  thinkingText?: string;
  errorMessage?: string;
  agent: string;
  model: ModelInfo;
  system?: string;
  tools?: Record<string, boolean>;
}

export interface AssistantMessage {
  id: string;
  sessionID: string;
  role: 'assistant';
  time: MessageTime;
  status: MessageStatus;
  thinkingText?: string;
  errorMessage?: string;
  parentID: string;
  modelID: string;
  providerID: string;
  mode: string;
  path: PathInfo;
  cost: number;
  tokens: TokenInfo;
  finish?: string;
  summary?: boolean;
  error?: Record<string, unknown>;
}

export type Message = UserMessage | AssistantMessage;

export function isUserMessage(msg: Message): msg is UserMessage {
  return msg.role === 'user';
}

export function isAssistantMessage(msg: Message): msg is AssistantMessage {
  return msg.role === 'assistant';
}
